### ISSUE TITLE
Release SecureDrop 1.4.0

### ISSUE THREAD

**Issue by conorsch** [freedomofpress/securedrop#5289]
This is a tracking issue for the release of SecureDrop 1.4.0

**String and feature freeze:** N/A (no string changes)
**String comment period:**  N/A (no string changes)
**Translation period:** N/A (no string changes)
**Pre-release announcement:** 2020-06-10
**Translation freeze:** N/A (no string changes)
**Release date:** 2020-06-17

**Release manager:** @conorsch 
**Deputy release manager:** @creviera 
**Localization manager**: N/A
**Deputy localization manager**: N/A

 
_SecureDrop maintainers and testers:_ As you QA 1.4.0, please report back your testing results as comments on this ticket. File GitHub issues for any problems found, tag them "QA: Release", and associate them with the [1.4.0 milestone](https://github.com/freedomofpress/securedrop/milestone/62) for tracking (or ask a maintainer to do so). The test plan TK

Test debian packages will be posted on https://apt-test.freedom.press signed with [the test key](https://gist.githubusercontent.com/conorsch/ec4008b111bc3142fca522693f3cce7e/raw/2968621e8ad92db4505a31fcc5776422d7d26729/apt-test%2520apt%2520pubkey). An Ansible playbook testing the upgrade path is [here](https://gist.github.com/rmol/eb605e28e290e90f45347f3a2f1ad43e).

# [QA Matrix for 1.4.0](https://docs.google.com/spreadsheets/d/15qr2XMlBrlCwgaErGzW9HR4UXJsRtwadwkhOPpEJmec/edit#gid=0) 
# [Test Plan for 1.4.0](https://github.com/freedomofpress/securedrop/wiki/1.4.0-Test-plan) 

# Prepare release candidate (1.4.0~rc1)
- [x] Prepare 1.4.0~rc1 release changelog 
- [x] Branch off release/1.4.0 from develop
- [x] Prepare 1.4.0~rc1
- [x] Build debs and put up `1.4.0~rc1` on test apt server (this is now done via a PR into [this repository](https://github.com/freedomofpress/securedrop-dev-packages-lfs)) - https://github.com/freedomofpress/securedrop-dev-packages-lfs/pull/44
- [x] Commit build logs to https://github.com/freedomofpress/build-logs 

After each test, please update the QA matrix and post details for Basic Server Testing, Application Acceptance Testing and 1.4.0-specific testing below in comments to this ticket.

# Final release
- [x] Ensure builder in release branch is updated and/or update builder image  
- [x] Push signed tag 
- [x] Build final Debian packages for 1.4.0 (and preserve build logs)
- [x] Commit package build logs to https://github.com/freedomofpress/build-logs
- [x] Upload Debian packages to apt QA server (including new tor packages)
- [x] Pre-Flight: Test install and upgrade (both cron-apt on Xenial, and Ansible on Xenial) of 1.4.0 works w/ prod repo debs, test updater logic in Tails
- [x] Flip apt QA server to prod status
- [x] merge the release branch changes to ``master`` and verify new docs build
- [x] Prepare and distribute release messaging 
- [x] Update release key on keyservers
- [x] Update release key on securedrop.org

# Post release
- [x] Create GitHub release object
- [x] Merge changelog back to `develop`
- [x] Update upgrade testing boxes
- [x] Update roadmap wiki page

**Comment by creviera** [comment]
We'll need to correct [the 1.4.0 changelog](https://github.com/freedomofpress/securedrop/commit/efd1121e14805e324bd719c3fc51fb69805cb960) by replaceing `#2527` with `#5257`

**Comment by creviera** [comment]
https://github.com/freedomofpress/securedrop/wiki/1.4.0-Test-Plan is now live. Next, we need to update the hagrid keyserver and securedrop.org webiste with the new release signing key.

**Comment by creviera** [comment]
https://docs.google.com/spreadsheets/d/15qr2XMlBrlCwgaErGzW9HR4UXJsRtwadwkhOPpEJmec/edit#gid=0 is ready (hardware columns we can skip during QA are marked `[SKIP]`) and release-specific rows are at the bottom of the matrix:
```
1.4.0-specific Version Update Testing: tor, ossec, and keyring pkgs  (details in issue)
1.4.0-specificWeb Application Testing: source_deleter (details in issue)
1.4.0-specificAdmin Workstation Testing: iptables check(details in issue)
1.4.0-specificAdmin Workstation Testing: multiple nameservers(details in issue)
1.4.0-specificAdmin Workstation Testing: keyring with new expire date(details in issue)
```

**Comment by kushaldas** [comment]
## 1.4.0 QA Checklist
 
For both upgrades and fresh installs, here is a list of functionality that requires testing. You can use this for copy/pasting into your QA report. Feel free to edit this message to update the plan as appropriate.
 
If you have submitted a QA report already for a 1.4.0 release candidate with successful [[Basic Server Testing]] and [[Application Acceptance Testing]], then you can skip these sections in subsequent reports, unless otherwise indicated by the Release Manager. This is to ensure that you focus your QA effort on the 1.4.0-specific changes as well as changes since the previous release candidate.
 
### Environment
 
- Install target: prod vm
- Tails version: 4.7
- Test Scenario: update
- SSH over Tor: no
- Onion service version: 2 and 3
- Release candidate: rc1
- General notes:
 
### Basic Server Testing
 
- [x] I can access both the source and journalist interfaces
- [x] I can SSH into both machines over Tor
- [x] AppArmor is loaded on app
    - [x] 0 processes are running unconfined
- [x] AppArmor is loaded on mon
    - [x] 0 processes are running unconfined
- [x] Both servers are running grsec kernels
- [x] iptables rules loaded
- [x] OSSEC emails begin to flow after install
- [x] OSSEC emails are decrypted to correct key and I am able to decrypt them
- [ ] [QA Matrix](https://docs.google.com/spreadsheets/d/15qr2XMlBrlCwgaErGzW9HR4UXJsRtwadwkhOPpEJmec) checks pass
 
#### Command Line User Generation
 
- [x] Can successfully add admin user and login
 
#### Administration
 
- [x] I have backed up and successfully restored the app server following [the backup documentation](https://docs.securedrop.org/en/latest/backup_and_restore.html)
- [x] If doing upgrade testing, make a backup on 1.3.0 and restore this backup on 1.4.0
- [x] "Send Test OSSEC Alert" button in the journalist triggers an OSSEC alert and an email is sent
- [x] Can successfully add journalist account with HOTP authentication 
 
### Application Acceptance Testing
 
#### Source Interface
 
##### Landing page base cases
- [x] JS warning bar does not appear when using Security Slider high
- [x] JS warning bar does appear when using Security Slider Low
 
##### First submission base cases
- [x] On generate page, refreshing codename produces a new 7-word codename
- [x] On submit page, empty submissions produce flashed message
- [x] On submit page, short message submitted successfully
- [x] On submit page, file greater than 500 MB produces "The connection was reset" in Tor Browser _quickly_ before the entire file is uploaded
- [x] On submit page, file less than 500 MB submitted successfully
 
##### Returning source base cases
 
- [x] Nonexistent codename cannot log in
- [x] Empty codename cannot log in
- [x] Legitimate codename can log in
- [x] Returning user can view journalist replies - need to log into journalist interface to test
 
#### Journalist Interface
 
##### Login base cases
- [x] Can log in with 2FA tokens
- [x] incorrect password cannot log in
- [x] invalid 2fa token cannot log in
- [x] 2fa immediate reuse cannot log in
- [x] Journalist account with HOTP can log in
 
##### Index base cases

- [x] Filter by codename works
- [x] Starring and unstarring works
- [x] Click select all selects all submissions
- [x] Selecting all and clicking "Download" works
 
##### Individual source page

- [x] You can submit a reply and a flashed message and new row appears
- [x] You cannot submit an empty reply
- [x] Clicking "Delete Source And Submissions" and the source and docs are deleted
- [x] You can click on a document and successfully decrypt using application private key
 
### Basic Tails Testing
 
#### Updater GUI
 
After updating to this release candidate and running `securedrop-admin tailsconfig`
 
- [x] The Updater GUI appears on boot
- [x] Updating occurs without issue
 
### 1.4.0: Release-specific changes

#### Version updates

##### [#5292](https://github.com/freedomofpress/securedrop/pull/5292)
 - [x] Verify that Tor 0.4.3.5 is running by running `sudo apt-cache policy tor`
##### [#5277](https://github.com/freedomofpress/securedrop/pull/5277)
 - [x] Verify that the new `securedrop-keyring` package has an updated expiration date of **2021-06-30** by running `sudo apt-key finger`
##### [#5287](https://github.com/freedomofpress/securedrop/pull/5287)
 - [x] Verify that the 1.4.0 `securedrop-ossec-agent` and `securedrop-ossec-server` packages are being used running `sudo apt-cache policy securedrop-ossec-agent` and `sudo apt-cache policy securedrop-ossec-server`.

#### Web Application

##### [#5257](https://github.com/freedomofpress/securedrop/pull/5257)
 - [x] Verify the `source_deleter` service is running on the app server
   - Use `qa_loader.py` or `create-dev-data.py` to populate the staging database with a large number of sources (100 should be sufficient to have triggered the timeout before this change, but 500 or 1000 would be better).
   - While the database is being populated, run `systemctl status securedrop_source_deleter`
 - [x] Verify that the JI delete action responds quickly (no longer times out) and sources are no longer available.
   - Once the database is populated, visit the journalist interface and select all but a few of the sources for deletion.
 - [x] Verify that `source_deleter` deletes all sources (Deleting 1000 sources took around a half an hour during PR testing)
   - Monitor the `source_deleter` logs on the app server with `journalctl -fu securedrop_source_deleter`
 - [x] Verify that sources not deleted are still present once deletion is complete

#### Admin Workstation

##### [#5287](https://github.com/freedomofpress/securedrop/pull/5287)
 - [ ] Verify that you receive an OSSEC alert level 12 email that says “The iptables default drop rules are incorrect.” when iptables is broken.
   - [Upgrade scenario] Run `sdconfig` on 1.3.0 with a broken config and run `make install` (to repro known-bad state), then upgrade via `apt-test`.
   - Break the `/etc/networking/iptables/rules_ipv4` file on the servers after installation (by substituting invalid values into fields, for example), then reboot the servers.

##### [#5288](https://github.com/freedomofpress/securedrop/pull/5288)
 - [ ] Verify that there are no errors in the sdconfig logic when multiple nameservers are provided during `sdconfig`
   - On an admin workstation, run `securedrop-admin sdconfig`. Try giving a single nameserver, then a list separated by commas, then by spaces. Enter invalid IP addresses. Try to enter more than three nameservers.
- [ ] Verify that the `dns_server` value in `install_files/ansible-base/group_vars/all/site-specific` is a list, not a string after specifying multiple nameservers.
- [ ] Verify that `securedrop-admin install` completes successfully and that each server's `/etc/resolvconf/resolv.conf.d/base` contains the correct nameservers.
- [ ] Verify that the dns rules contain the nameservers you provided by running Run `sudo iptables -S | grep`
- [ ] Verify that the value is correctly transformed to a list and that the server's `/etc/resolvconf/resolv.conf.d/base` and iptables rules are correct
   - Edit `install_files/ansible-base/group_vars/all/site-specific` to make `dns_server` a string, covering at least some of the variations you tested above: single, comma-separated, space-separated. Upon each change, run `securedrop-admin install`. 

##### [#5277](https://github.com/freedomofpress/securedrop/pull/5277)
- Since we can't rely on the updater to get the new key until the release object is published (see https://github.com/zenmonkeykstop/securedrop/blob/ee9b16e39908b1bac9d9aba202eb9a50193a0923/admin/securedrop_admin/__init__.py#L665-L667), we will need to create temporary keyrings for each check.

   - [ ] Verify the SecureDrop Release signing key is present with the new expiration date of **2021-06-30** on the keyserver
      - `mkdir -m 700 /tmp/qa-new-key-on-keyserver`
      - `gpg --homedir /tmp/qa-new-key-on-keyserver --keyserver hkps://keys.openpgp.org --recv-keys 22245C81E3BAEB4138B36061310F561200F4AD77`
      - `gpg --homedir /tmp/qa-new-key-on-keyserver -k`

   - [ ] Verify the SecureDrop Release signing key is present with the new expiration date of **2021-06-30** on securedrop.org
      - `mkdir -m 700 /tmp/qa-new-key-on-website`
      - `curl -LO https://securedrop.org/securedrop-release-key.asc`
      - `gpg --homedir /tmp/qa-new-key-on-website --import securedrop-release-key.asc`
      - `gpg --homedir /tmp/qa-new-key-on-website -k`

### Preflight 
 
- [ ] Ensure the builder image is up-to-date on release day
 
These tests should be performed the day of release prior to live debian packages on `apt.freedom.press`.
 
#### Basic testing
 
- [ ] Install or upgrade occurs without error
- [ ] Source interface is available and version string indicates it is 1.4.0
- [ ] A message can be successfully submitted 
 
#### Tails
 
- [ ] The updater GUI appears on boot
- [ ] The update successfully occurs to 1.4.0
- [ ] After reboot, updater GUI no longer appears

**Comment by rmol** [comment]
### Environment

- Install target: NUC 7i5BNH
- Tails version: 4.7
- Test Scenario: clean install
- SSH over Tor: yes
- Onion service version: 3
- Release candidate: 1.4.0rc1
- General notes:

### Basic Server Testing

- [x] I can access both the source and journalist interfaces
- [x] I can SSH into both machines over Tor
- [x] AppArmor is loaded on app
    - [x] 0 processes are running unconfined
- [x] AppArmor is loaded on mon
    - [x] 0 processes are running unconfined
- [x] Both servers are running grsec kernels
- [x] iptables rules loaded
- [x] OSSEC emails begin to flow after install
- [x] OSSEC emails are decrypted to correct key and I am able to decrypt them
- [x] [QA Matrix](https://docs.google.com/spreadsheets/d/15qr2XMlBrlCwgaErGzW9HR4UXJsRtwadwkhOPpEJmec) checks pass

#### Command Line User Generation

- [x] Can successfully add admin user and login

#### Administration

- [x] I have backed up and successfully restored the app server following [the backup documentation](https://docs.securedrop.org/en/latest/backup_and_restore.html)
- [ ] If doing upgrade testing, make a backup on 1.3.0 and restore this backup on 1.4.0
- [x] "Send Test OSSEC Alert" button in the journalist triggers an OSSEC alert and an email is sent
- [x] Can successfully add journalist account with HOTP authentication

### Application Acceptance Testing

#### Source Interface

##### Landing page base cases
- [x] JS warning bar does not appear when using Security Slider high
- [x] JS warning bar does appear when using Security Slider Low

##### First submission base cases
- [x] On generate page, refreshing codename produces a new 7-word codename
- [x] On submit page, empty submissions produce flashed message
- [x] On submit page, short message submitted successfully
- [x] On submit page, file greater than 500 MB produces "The connection was reset" in Tor Browser _quickly_ before the entire file is uploaded
- [x] On submit page, file less than 500 MB submitted successfully

##### Returning source base cases

- [x] Nonexistent codename cannot log in
- [x] Empty codename cannot log in
- [x] Legitimate codename can log in
- [x] Returning user can view journalist replies - need to log into journalist interface to test

#### Journalist Interface

##### Login base cases
- [x] Can log in with 2FA tokens
- [x] incorrect password cannot log in
- [x] invalid 2fa token cannot log in
- [x] 2fa immediate reuse cannot log in
- [x] Journalist account with HOTP can log in

##### Index base cases

- [x] Filter by codename works
- [x] Starring and unstarring works
- [x] Click select all selects all submissions
- [x] Selecting all and clicking "Download" works

##### Individual source page

- [x] You can submit a reply and a flashed message and new row appears
- [x] You cannot submit an empty reply
- [x] Clicking "Delete Source And Submissions" and the source and docs are deleted
- [x] You can click on a document and successfully decrypt using application private key

### Basic Tails Testing

#### Updater GUI

After updating to this release candidate and running `securedrop-admin tailsconfig`

- [x] The Updater GUI appears on boot
- [x] Updating occurs without issue

### 1.4.0: Release-specific changes

#### Version updates

##### [#5292](https://github.com/freedomofpress/securedrop/pull/5292)
 - [x] Verify that Tor 0.4.3.5 is running by running `sudo apt-cache policy tor`
##### [#5277](https://github.com/freedomofpress/securedrop/pull/5277)
 - [x] Verify that the new `securedrop-keyring` package has an updated expiration date of **2021-06-30** by running `sudo apt-key finger`
##### [#5287](https://github.com/freedomofpress/securedrop/pull/5287)
 - [x] Verify that the 1.4.0 `securedrop-ossec-agent` and `securedrop-ossec-server` packages are being used running `sudo apt-cache policy securedrop-ossec-agent` and `sudo apt-cache policy securedrop-ossec-server`.

#### Web Application

##### [#5257](https://github.com/freedomofpress/securedrop/pull/5257)
 - [x] Verify the `source_deleter` service is running on the app server
   - Use `qa_loader.py` or `create-dev-data.py` to populate the staging database with a large number of sources (100 should be sufficient to have triggered the timeout before this change, but 500 or 1000 would be better).
   - While the database is being populated, run `systemctl status securedrop_source_deleter`
 - [x] Verify that the JI delete action responds quickly (no longer times out) and sources are no longer available.
   - Once the database is populated, visit the journalist interface and select all but a few of the sources for deletion.
 - [x] Verify that `source_deleter` deletes all sources (Deleting 1000 sources took around a half an hour during PR testing)
   - Monitor the `source_deleter` logs on the app server with `journalctl -fu securedrop_source_deleter`
 - [x] Verify that sources not deleted are still present once deletion is complete

#### Admin Workstation

##### [#5287](https://github.com/freedomofpress/securedrop/pull/5287)
 - [x] Verify that you receive an OSSEC alert level 12 email that says “The iptables default drop rules are incorrect.” when iptables is broken.
   - [Upgrade scenario] Run `sdconfig` on 1.3.0 with a broken config and run `make install` (to repro known-bad state), then upgrade via `apt-test`.
   - Break the `/etc/networking/iptables/rules_ipv4` file on the servers after installation (by substituting invalid values into fields, for example), then reboot the servers.

##### [#5288](https://github.com/freedomofpress/securedrop/pull/5288)
 - [x] Verify that there are no errors in the sdconfig logic when multiple nameservers are provided during `sdconfig`
   - On an admin workstation, run `securedrop-admin sdconfig`. Try giving a single nameserver, then a list separated by commas, then by spaces. Enter invalid IP addresses. Try to enter more than three nameservers.
- [x] Verify that the `dns_server` value in `install_files/ansible-base/group_vars/all/site-specific` is a list, not a string after specifying multiple nameservers.
- [x] Verify that `securedrop-admin install` completes successfully and that each server's `/etc/resolvconf/resolv.conf.d/base` contains the correct nameservers.
- [x] Verify that the dns rules contain the nameservers you provided by running Run `sudo iptables -S | grep`
- [x] Verify that the value is correctly transformed to a list and that the server's `/etc/resolvconf/resolv.conf.d/base` and iptables rules are correct
   - Edit `install_files/ansible-base/group_vars/all/site-specific` to make `dns_server` a string, covering at least some of the variations you tested above: single, comma-separated, space-separated. Upon each change, run `securedrop-admin install`.

##### [#5277](https://github.com/freedomofpress/securedrop/pull/5277)
- Since we can't rely on the updater to get the new key until the release object is published (see https://github.com/zenmonkeykstop/securedrop/blob/ee9b16e39908b1bac9d9aba202eb9a50193a0923/admin/securedrop_admin/__init__.py#L665-L667), we will need to create temporary keyrings for each check.

   - [x] Verify the SecureDrop Release signing key is present with the new expiration date of **2021-06-30** on the keyserver
      - `mkdir -m 700 /tmp/qa-new-key-on-keyserver`
      - `gpg --homedir /tmp/qa-new-key-on-keyserver --keyserver hkps://keys.openpgp.org --recv-keys 22245C81E3BAEB4138B36061310F561200F4AD77`
      - `gpg --homedir /tmp/qa-new-key-on-keyserver -k`

   - [x] Verify the SecureDrop Release signing key is present with the new expiration date of **2021-06-30** on securedrop.org
      - `mkdir -m 700 /tmp/qa-new-key-on-website`
      - `curl -LO https://securedrop.org/securedrop-release-key.asc`
      - `gpg --homedir /tmp/qa-new-key-on-website --import securedrop-release-key.asc`
      - `gpg --homedir /tmp/qa-new-key-on-website -k`

**Comment by zenmonkeykstop** [comment]
## QA plan
 
There is no kernel update in this release, so we’ll be testing on fewer hardware platforms. However, given the importance of the iptables check, we’ll test the officially supported hardware:

- NUC7
- 1U test server
 
## 1.4.0 QA Checklist
 
For both upgrades and fresh installs, here is a list of functionality that requires testing. You can use this for copy/pasting into your QA report. Feel free to edit this message to update the plan as appropriate.
 
If you have submitted a QA report already for a 1.4.0 release candidate with successful [[Basic Server Testing]] and [[Application Acceptance Testing]], then you can skip these sections in subsequent reports, unless otherwise indicated by the Release Manager. This is to ensure that you focus your QA effort on the 1.4.0-specific changes as well as changes since the previous release candidate.
 
### Environment
 
- Install target: 1U servers
- Tails version: 4.7
- Test Scenario: cron-apt update
- SSH over Tor: yes
- Onion service version: v3
- Release candidate: 1.4.0-rc1
- General notes:
 
### Basic Server Testing
 
- [x] I can access both the source and journalist interfaces
- [x] I can SSH into both machines over Tor
- [x] AppArmor is loaded on app
    - [x] 0 processes are running unconfined
- [x] AppArmor is loaded on mon
    - [x] 0 processes are running unconfined
- [x] Both servers are running grsec kernels
- [ ] iptables rules loaded **FAIL, as set up with 1.3.0 broken dns field**
- [x] OSSEC emails begin to flow after install
- [x] OSSEC emails are decrypted to correct key and I am able to decrypt them
- [x] [QA Matrix](https://docs.google.com/spreadsheets/d/15qr2XMlBrlCwgaErGzW9HR4UXJsRtwadwkhOPpEJmec) checks pass
 
#### Command Line User Generation
 
- [x] Can successfully add admin user and login
 
#### Administration
 
- [x] I have backed up and successfully restored the app server following [the backup documentation](https://docs.securedrop.org/en/latest/backup_and_restore.html)
- [ ] If doing upgrade testing, make a backup on 1.3.0 and restore this backup on 1.4.0 **SKIPPED. no backup available**
- [ ] "Send Test OSSEC Alert" button in the journalist triggers an OSSEC alert and an email is sent **FAIL - alert not sent, nothing in logs**
- [ ] Can successfully add journalist account with HOTP authentication **Not tested**
 
### Application Acceptance Testing
 
#### Source Interface
 
##### Landing page base cases
- [x] JS warning bar does not appear when using Security Slider high
- [x] JS warning bar does appear when using Security Slider Low
 
##### First submission base cases
- [x] On generate page, refreshing codename produces a new 7-word codename
- [x] On submit page, empty submissions produce flashed message
- [x] On submit page, short message submitted successfully
- [x] On submit page, file greater than 500 MB produces "The connection was reset" in Tor Browser _quickly_ before the entire file is uploaded
- [x] On submit page, file less than 500 MB submitted successfully
 
##### Returning source base cases
 
- [x] Nonexistent codename cannot log in
- [x] Empty codename cannot log in
- [x] Legitimate codename can log in
- [x] Returning user can view journalist replies - need to log into journalist interface to test
 
#### Journalist Interface
 
##### Login base cases
- [x] Can log in with 2FA tokens
- [x] incorrect password cannot log in
- [x] invalid 2fa token cannot log in
- [x] 2fa immediate reuse cannot log in
- [ ] Journalist account with HOTP can log in **not tested**
 
##### Index base cases

- [x] Filter by codename works
- [x] Starring and unstarring works
- [x] Click select all selects all submissions
- [x] Selecting all and clicking "Download" works
 
##### Individual source page

- [x] You can submit a reply and a flashed message and new row appears
- [x] You cannot submit an empty reply
- [x] Clicking "Delete Source And Submissions" and the source and docs are deleted
- [x] You can click on a document and successfully decrypt using application private key
 
### Basic Tails Testing
 
#### Updater GUI
 
After updating to this release candidate and running `securedrop-admin tailsconfig`
 
- [x] The Updater GUI appears on boot
- [x] Updating occurs without issue
 
### 1.4.0: Release-specific changes

#### Version updates

##### [#5292](https://github.com/freedomofpress/securedrop/pull/5292)
 - [x] Verify that Tor 0.4.3.5 is running by running `sudo apt-cache policy tor`
##### [#5277](https://github.com/freedomofpress/securedrop/pull/5277)
 - [x] Verify that the new `securedrop-keyring` package has an updated expiration date of **June 30, 2021** by running `sudo apt-key finger`
##### [#5287](https://github.com/freedomofpress/securedrop/pull/5287)
 - [x] Verify that the 1.4.0 `securedrop-ossec-agent` and `securedrop-ossec-server` packages are being used running `sudo apt-cache policy securedrop-ossec-agent` and `sudo apt-cache policy securedrop-ossec-server`.

#### Web Application

##### [#5257](https://github.com/freedomofpress/securedrop/pull/5257)
 - [x] Verify the `source_deleter` service is running on the app server
   - Use `qa_loader.py` or `create-dev-data.py` to populate the staging database with a large number of sources (100 should be sufficient to have triggered the timeout before this change, but 500 or 1000 would be better).
   - While the database is being populated, run `systemctl status securedrop_source_deleter`
 - [x] Verify that the JI delete action responds quickly (no longer times out) and sources are no longer available.
   - Once the database is populated, visit the journalist interface and select all but a few of the sources for deletion.
 - [x] Verify that `source_deleter` deletes all sources (Deleting 1000 sources took around a half an hour during PR testing)
   - Monitor the `source_deleter` logs on the app server with `journalctl -fu securedrop_source_deleter`
 - [x] Verify that sources not deleted are still present once deletion is complete

#### Admin Workstation

##### [#5287](https://github.com/freedomofpress/securedrop/pull/5287)
 - [x] Verify that you receive an OSSEC alert level 12 email that says “The iptables default drop rules are incorrect.” when iptables is broken.
   - [Upgrade scenario] Run `sdconfig` on 1.3.0 with a broken config and run `make install` (to repro known-bad state), then upgrade via `apt-test`.
   - Break the `/etc/networking/iptables/rules_ipv4` file on the servers after installation (by substituting invalid values into fields, for example), then reboot the servers.

##### [#5288](https://github.com/freedomofpress/securedrop/pull/5288)
 - [x] Verify that there are no errors in the sdconfig logic when multiple nameservers are provided during `sdconfig`
   - On an admin workstation, run `securedrop-admin sdconfig`. Try giving a single nameserver, then a list separated by commas, then by spaces. Enter invalid IP addresses. Try to enter more than three nameservers.
- [x] Verify that the `dns_server` value in `install_files/ansible-base/group_vars/all/site-specific` is a list, not a string after specifying multiple nameservers.
- [x] Verify that `securedrop-admin install` completes successfully and that each server's `/etc/resolvconf/resolv.conf.d/base` contains the correct nameservers. **tested with single entry: pass. tested with double entry: pass**
- [x] Verify that the dns rules contain the nameservers you provided by running Run `sudo iptables -S | grep`
- [x] Verify that the value is correctly transformed to a list and that the server's `/etc/resolvconf/resolv.conf.d/base` and iptables rules are correct
   - Edit `install_files/ansible-base/group_vars/all/site-specific` to make `dns_server` a string, covering at least some of the variations you tested above: single, comma-separated, space-separated. Upon each change, run `securedrop-admin install`. 

##### [#5277](https://github.com/freedomofpress/securedrop/pull/5277)
- [x] Verify the SecureDrop Release signing key is present with the new expiration date of **June 30, 2021** in the local keyring.
   - After install, run `gpg -k securedrop-release-key@freedom.press`.

### Preflight 
 
- [ ] Ensure the builder image is up-to-date on release day
 
These tests should be performed the day of release prior to live debian packages on `apt.freedom.press`.
 
#### Basic testing
 
- [ ] Install or upgrade occurs without error
- [ ] Source interface is available and version string indicates it is 1.4.0
- [ ] A message can be successfully submitted 
 
#### Tails
 
- [ ] The updater GUI appears on boot
- [ ] The update successfully occurs to 1.4.0
- [ ] After reboot, updater GUI no longer appears

**Comment by conorsch** [comment]
> "Send Test OSSEC Alert" button in the journalist triggers an OSSEC alert and an email is sent **FAIL - alert not sent, nothing in logs**

Sounds like #5265. Failure may be logged under procmail. 

>  Verify that you receive an OSSEC alert level 12 email that says “The iptables default drop rules are incorrect.” when iptables is broken. **FAIL: got OSSEC alert after initial (natural) cron-apt, but no alerts of any kind after running playbook to correct DNS server entries**

I _do_ see such an email alert from 2020-06-11, but that's likely too old for the current test report you're filing. What about in `/var/ossec/logs/alerts/alerts.log`, do you see a level 12 alert there?

**Comment by eloquence** [comment]
> Sounds like #5265.

That was my original experience after upgrading to 1.3.0, before rebooting. After rebooting, the test alerts did send (while permission errors continued to be logged), with the email body documented here:

https://github.com/freedomofpress/securedrop/issues/5265#issuecomment-628869882

**Comment by conorsch** [comment]
All PRs have been backported, proceeding with final tag for 1.4.0.

**Comment by zenmonkeykstop** [comment]
(For historical purposes, the OSSEC agent failure mentioned above was tracked down to a misconfiguration in QA firewall rules, not a code issue)

**Comment by eloquence** [comment]
Fully closed out by #5329.
